tools/biopattern.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/python
+# @lint-avoid-python-3-compatibility-imports
+#
+# biopattern - Identify random/sequential disk access patterns.
+#              For Linux, uses BCC, eBPF.
+#
+# Copyright (c) 2022 rockyxing.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 21-Feb-2022   rockyxing   Created this.
+
+from __future__ import print_function
+from bcc import BPF
+from time import sleep, strftime
+import argparse
+import os
+
+examples = """examples:
+    ./biopattern            # show block device I/O pattern.
+    ./biopattern 1 10       # print 1 second summaries, 10 times
+    ./biopattern -d sdb     # show sdb only
+"""
+parser = argparse.ArgumentParser(
+    description="Show block device I/O pattern.",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-d", "--disk", type=str,
+    help="Trace this disk only")
+parser.add_argument("interval", nargs="?", default=99999999,
+    help="Output interval in seconds")
+parser.add_argument("count", nargs="?", default=99999999,
+    help="Number of outputs")
+args = parser.parse_args()
+countdown = int(args.count)
+
+bpf_text="""
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+
+struct counter {
+    u64 last_sector;
+    u64 bytes;
+    u32 sequential;
+    u32 random;
+};
+
+BPF_HASH(counters, u32, struct counter);
+
+TRACEPOINT_PROBE(block, block_rq_complete)
+{
+    struct counter *counterp;
+    struct counter zero = {};
+    u32 dev = args->dev;
+    u64 sector = args->sector;
+    u32 nr_sector = args->nr_sector;
+
+    DISK_FILTER
+
+    counterp = counters.lookup_or_try_init(&dev, &zero);
+    if (counterp == 0) {
+        return 0;
+    }
+
+    if (counterp->last_sector) {
+        if (counterp->last_sector == sector) {
+            __sync_fetch_and_add(&counterp->sequential, 1);
+        } else {
+            __sync_fetch_and_add(&counterp->random, 1);
+        }
+        __sync_fetch_and_add(&counterp->bytes, nr_sector * 512);
```

Comment:
I think this line should move out of if block.

Reply:
if counterp->last_sector equals 0, means counterp not found and inited。First time you can not determine whether this I/O is seq or random，so here add seq or random at next time（counterp->last_sector > 0）.